pkg/observer/polling/observer.go

```diff
@@ -0,0 +1,337 @@
+}
+
+// HeadProvider is a dependency for a block data source channel
+type HeadProvider interface {
+	HeadTicker() chan ocr2keepers.BlockKey
+}
+
+// Service ...
+type Service interface {
+	Start()
+	Stop()
+}
+
+// Shuffler ...
+type Shuffler[T any] interface {
+	Shuffle([]T) []T
+}
+
+// Ratio is an interface that provides functions to calculate a ratio of a given
+// input
+type Ratio interface {
+	// OfInt should return n out of x such that n/x ~ r (ratio)
+	OfInt(int) int
+	fmt.Stringer
+}
+
+// NewPollingObserver ...
+func NewPollingObserver(
+	logger *log.Logger,
+	src UpkeepProvider,
+	heads HeadProvider,
+	runner Runner,
+	encoder Encoder,
+	ratio Ratio,
+	maxSamplingDuration time.Duration, // maximum amount of time allowed for RPC calls per head
+	coord ocr2keepers.Coordinator,
+	mercuryLookup bool,
+) *PollingObserver {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ob := &PollingObserver{
+		ctx:              ctx,
+		cancel:           cancel,
+		logger:           logger,
+		samplingDuration: maxSamplingDuration,
+		shuffler:         util.Shuffler[ocr2keepers.UpkeepKey]{Source: util.NewCryptoRandSource()}, // use crypto/rand shuffling for true random
+		ratio:            ratio,
+		stager:           &stager{},
+		coordinator:      coord,
+		src:              src,
+		runner:           runner,
+		encoder:          encoder,
+		heads:            heads,
+		mercuryLookup:    mercuryLookup,
+	}
+
+	// make all go-routines started by this entity automatically recoverable
+	// on panics
+	ob.services = []Service{
+		util.NewRecoverableService(&observer.SimpleService{F: ob.runHeadTasks, C: cancel}, logger),
+	}
+
+	return ob
+}
+
+type PollingObserver struct {
+	ctx       context.Context
+	cancel    context.CancelFunc
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// static values provided by constructor
+	samplingDuration time.Duration // limits time spent processing a single block
+	ratio            Ratio         // ratio for limiting sample size
+
+	// initialized components inside a constructor
+	services []Service
+	stager   *stager
+
+	// dependency interfaces required by the polling observer
+	logger      *log.Logger
+	heads       HeadProvider                    // provides new blocks to be operated on
+	coordinator ocr2keepers.Coordinator         // key status coordinator tracks in-flight status
+	shuffler    Shuffler[ocr2keepers.UpkeepKey] // provides shuffling logic for upkeep keys
+
+	src           UpkeepProvider
+	runner        Runner
+	encoder       Encoder
+	mercuryLookup bool
+}
+
+// Observe implements the Observer interface and provides a slice of identifiers
+// that were observed to be performable along with the block at which they were
+// observed. All ids that are pending are filtered out.
+func (o *PollingObserver) Observe() (ocr2keepers.BlockKey, []ocr2keepers.UpkeepIdentifier, error) {
+	bl, ids := o.stager.get()
```

Comment:
is this new component being used in 2.0 or just new for 2.1?

Reply:
This is used in 2.0, not in 2.1